File: benchmarks/typescript/parse_user.ts
type User = {
  name: string;
  age: number;
};

type ParseResult =
  | { ok: true; value: User }
  | { ok: false; error: string };

export function parseUser(input: unknown): ParseResult {
  if (typeof input !== "object" || input === null) {
    return { ok: false, error: "expected object" };
  }
  const obj = input as Record<string, unknown>;
  if (typeof obj.name !== "string") {
    return { ok: false, error: "name: expected string" };
  }
  if (typeof obj.age !== "number") {
    return { ok: false, error: "age: expected number" };
  }
  return { ok: true, value: { name: obj.name, age: obj.age } };
}

File: benchmarks/go/parse_user.go
package parseuser

import "errors"

type User struct {
	Name string
	Age  int
}

func ParseUser(input interface{}) (User, error) {
	obj, ok := input.(map[string]interface{})
	if !ok {
		return User{}, errors.New("expected object")
	}
	name, ok := obj["name"].(string)
	if !ok {
		return User{}, errors.New("name: expected string")
	}
	age, ok := obj["age"].(float64)
	if !ok {
		return User{}, errors.New("age: expected number")
	}
	return User{Name: name, Age: int(age)}, nil
}
